### PR TITLE
additional prefab transition processes

### PR DESCRIPTION
**Pull request by slwu89**
This PR adds functions `fixed_probability_multinomial_process`, `multi_probability_multinomial_process`, and `multi_probability_bernoulli_process` which address issues #40 and #39 , allowing for multinomial processes with constant and individual-specific probability of leaving the source compartment, and Bernoulli process with individual specific probability. The individual-specific probabilities are expected to be contained in a `DoubleVariable` object.

This PR also moves the logic in `bitset_sample` (in bitset.cpp) to `bitset_sample_internal` in IterableBitset.h; this was needed because the prefab transitions in prefab.cpp needed to use that function's logic. It also has a new function `bitset_sample_multi_internal` in IterableBitset.h which samples a bitset with templated input iterators, necessary for `multi_probability_multinomial_process` to work correctly.

Tests for the prefab functions have also been added!

### CODE REVIEW

File: src/prefab.cpp [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+//' probabilities contained in the vector 'destination_probabilities'.
+//' @param variable a \code{\link{CategoricalVariable}} object
+//' @param source_state a string representing the source state
+//' @param destination_states a vector of strings representing the destination states
+//' @param rate probability of individuals in source state to leave
+//' @param destination_probabilities probability vector of destination states
+//' @export
+// [[Rcpp::export]]
+Rcpp::XPtr<process_t> fixed_probability_multinomial_process(
+    const Rcpp::Environment variable,
```
Comment (giovannic) [inline src/prefab.cpp:27]: Could we instead pass an Rcpp::XPtr<DoubleVariable>?

A few reasons:
 * It adds a bit of type checking
 * I have found incomprehensible errors when I accidentaly put in an R thing that is not what we expect and gets casted into an Rcpp::Environment

Then we could have a small function in R/prefab.R which checks the CategoricalVariable and passes .variable to this function

File: src/prefab.cpp [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+//' probabilities contained in the vector 'destination_probabilities'.
+//' @param variable a \code{\link{CategoricalVariable}} object
+//' @param source_state a string representing the source state
+//' @param destination_states a vector of strings representing the destination states
+//' @param rate_variable \code{\link{DoubleVariable}} giving individual probability of each individual in source state to leave
+//' @param destination_probabilities probability vector of destination states
+//' @export
+// [[Rcpp::export]]
+Rcpp::XPtr<process_t> multi_probability_multinomial_process(
+    const Rcpp::Environment variable,
```
Comment (giovannic) [inline src/prefab.cpp:90]: Same as above

File: src/prefab.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+//' individual probability specified by the \code{\link{DoubleVariable}}
+//' object 'rate_variable'.
+//' @param variable a \code{\link{CategoricalVariable}} object
+//' @param from a string representing the source state
+//' @param to a string representing the destination state
+//' @param rate_variable \code{\link{DoubleVariable}} giving individual probability of each individual in source state to leave
+//' @export
+// [[Rcpp::export]]
+Rcpp::XPtr<process_t> multi_probability_bernoulli_process(
+    const Rcpp::Environment variable,
```
Comment (giovannic) [inline src/prefab.cpp:156]: same as above

**Review by giovannic — CHANGES_REQUESTED** [review]
Very nice! Few small nit picks

File: src/prefab.cpp (hunk 1 above)
Comment (slwu89) [reply src/prefab.cpp]: Totally. I wasn't sure the best way to pass those data down to C++, this is much better than using Rcpp::Environment.

Fixed in https://github.com/mrc-ide/individual/pull/63/commits/a3cba8683426608a9d8d1969cd68456157b75ac5 and also made `bitset_sample_multi_internal` a little simpler.

**Review by giovannic — APPROVED** [review]
Great, thanks!

**Pull request merged**